### PR TITLE
Solutions to both challenges.

### PR DESCRIPTION
**Pull request by rewanthtammana**
Please review this PR. @ashu210890 @ankuriitk 

You can clone the master branch to see the complete code. In this repo the base branch is `default`.

### CODE REVIEW

File: test/question-2.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+// Imports required modules.
+const assert = require('assert');
+const findFriendsInRadius = require('../Question-2/findFriendsInRadius');
+
+describe('Question 2', function() {
+    const input = [
+        {"latitude": "16.257447", "longitude": "80.440844", "user_id": 9, "name": "Place 1"},
+        {"latitude": "16.473541", "longitude": "80.655077", "user_id": 19, "name": "Place 2"},
+        {"latitude": "16.246899", "longitude": "80.462816", "user_id": 8, "name": "Place 3"},
```
Comment (ankuriitk) [inline test/question-2.js:12]: no tests for negative lat/lng? How about north/south pole?

File: test/question-1.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        assert.deepEqual(findTriplet(arr, sum), false);
+    });
+});
+
+describe('Question 1', function() {
+    const arr = [12, 12, 12, 34, 90];
+    const sum = 24;
+
+    it('case 7', function() {
+        assert.deepEqual(findTriplet(arr, sum), false);
```
Comment (ankuriitk) [inline test/question-1.js]: a function's return value should be single type - so instead of returning false, returning an empty array would have been better

File: test/question-1.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        assert.deepEqual(findTriplet(arr, sum), [1, 3, 5]);
+    });
+});
+
+describe('Question 1', function() {
+    const arr = [0, 0, 0, -3, 3];
+    const sum = 0;
+
+    it('case 2', function() {
+        assert.deepEqual(findTriplet(arr, sum), [-3, 0, 3]);
```
Comment (ankuriitk) [inline test/question-1.js:23]: 0, 0, 0 is also a valid one...what will it take to output all such triplets?

File: Question-2/constant.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// This file contains all the constant values which are used by the main function.
+
+module.exports = {
+    "partyCoordinates": {
+        "latitude": "12.9611159",
+        "longitude" : "77.6362214"
+    },
+    "URL": "https://gist.github.com/ankuriitk/d8792e7365a3ed154f2d4e6affb91324/raw/ed905a06a7e9f7908feb06127404bd7bf2ec828e/friends.json"
```
Comment (ankuriitk) [inline Question-2/constant.js:8]: good that you took it out from code - many candidates miss this.

File: test/question-2.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    it('case 2', function() {
+        assert.deepEqual(findFriendsInRadius(input, radius, partyCoordinates), {
+            "8": "Place 3",
+            "9": "Place 1",
+            "19": "Place 2"
+        });
+    });
+
+    radius = -50;
```
Comment (ankuriitk) [inline test/question-2.js:40]: -50 is valid?

File: test/question-2.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    it('edge case 1', function() {
+        assert.deepEqual(findFriendsInRadius(input, radius, partyCoordinates), {
+
+        });
+    });
+
+    it('edge case 2', function() {
+        assert.deepEqual(findFriendsInRadius(input, radius, partyCoordinates1), {
+             "8": "Place 3"
```
Comment (ankuriitk) [inline test/question-2.js]: nice!

File: Question-2/findFriendsInRadius.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    radius = Math.abs(radius);
+
+    for (let index in input) {
+        const entity = input[index];
+
+        // Takes two sets of geographic coordinates in decimal degrees and produces
+        // distance along the great circle line.
+        const distance = greatCircle.distance(partyCoordinates["latitude"],
+                                            partyCoordinates["longitude"],
+                                            entity["latitude"],
```
Comment (ankuriitk) [inline Question-2/findFriendsInRadius.js]: indentation is off

File: test/question-1.js (hunk 2 above)
Comment (rewanthtammana) [reply test/question-1.js]: Fixed. Link to commit - https://github.com/rewanth1997/granular-coding-round/pull/1/commits/fd0ee207344c21ffb6584a830535cc9fdbe3b349

File: Question-2/findFriendsInRadius.js (hunk 7 above)
Comment (rewanthtammana) [reply Question-2/findFriendsInRadius.js]: Fixed. Link to commit - https://github.com/rewanth1997/granular-coding-round/pull/1/commits/c9ad0efc4493b10938b8a006b22d18e50453ebb9

File: test/question-2.js (hunk 5 above)
Comment (rewanthtammana) [reply test/question-2.js:40]: Yes, if the user is uses this code w.r.t his/her friends location. If you doesn't feel it right, simply deleting [L8](https://github.com/rewanth1997/granular-coding-round/pull/1/files#diff-f415a5643ce2644cf2af297d98f99516R8) and adding an additional check `if(distance <= radius && distance >= 0)` in [L22](https://github.com/rewanth1997/granular-coding-round/pull/1/files#diff-f415a5643ce2644cf2af297d98f99516R22) fixes it.

Comment (rewanthtammana) [reply test/question-2.js:40]: I was referring to `findFriendsInRadius.js` file.

File: test/question-1.js (hunk 3 above)
Comment (rewanthtammana) [reply test/question-1.js:23]: Instead of returning the triplet on its first find, we have to store all the possibilities in an array and return it after the end for loop.
```
$ git diff
diff --git a/Question-1/findTriplet.js b/Question-1/findTriplet.js
index f202d02..c7a9aaf 100644
--- a/Question-1/findTriplet.js
+++ b/Question-1/findTriplet.js
@@ -3,6 +3,7 @@ module.exports = function findTriplet(arr, sum)
 {
     const size = arr.length;
     let left, right;
+    let output = [];

     // Sorts the integer array.
     arr.sort((num1, num2) => {
@@ -15,7 +16,9 @@ module.exports = function findTriplet(arr, sum)
         while(left < right) {
             let tempSum = arr[i] + arr[left] + arr[right];
             if(tempSum == sum) {
-                return [arr[i], arr[left], arr[right]];
+                output.push([arr[i], arr[left], arr[right]]);
+                left++;
+                right--;
             }
             else if(tempSum < sum) {
                 left++;
@@ -26,5 +29,5 @@ module.exports = function findTriplet(arr, sum)
         }
     }

-    return [];
+    return output;
 }
```

`output.push([arr[i], arr[left], arr[right], "\n"]);` -> This has to be used if you want to prettify the output instead of `output.push([arr[i], arr[left], arr[right]]);`

File: test/question-2.js (hunk 1 above)
Comment (rewanthtammana) [reply test/question-2.js:12]: We can use "-16.257447" to south latitude and "-80.440844" for west longitude in [L9](https://github.com/rewanth1997/granular-coding-round/pull/1/files/01658018dfa61d9bcd6106c36ea312c610478ee6#diff-ce2201b603840a8854d184917100bac6R9). They just work the same that's why I ignored them.

Comment (rewanthtammana) [reply test/question-2.js:12]: Fixed. Link to commit - https://github.com/rewanth1997/granular-coding-round/pull/1/commits/61e36f91a515464a13627d23d299e7a59cac5393

**Comment by rewanthtammana** [comment]
Please review all the changes @ankuriitk

**Comment by ashu210890** [comment]
Looks good overall.
Nitpick: It's hard to figure out the purpose of different 'describe' blocks in tests. It's usually advisable to have a descriptive describe text or (If the describe text is becoming too long) have a comment describing the category of tests this block contains.

File: test/question-1.js (hunk 3 above)
Comment (ankuriitk) [reply test/question-1.js:23]: :+1:

File: test/question-2.js (hunk 5 above)
Comment (ankuriitk) [reply test/question-2.js:40]: cool

File: test/question-2.js (hunk 1 above)
Comment (ankuriitk) [reply test/question-2.js:12]: negative lat/lng is fine. But I still I want to see a test where distance needs to be calulcated between north pole and south pole.

**Comment by rewanthtammana** [comment]
I added comments which give a basic idea on each `describe` block @ashu210890 . Link to commit - https://github.com/rewanth1997/granular-coding-round/pull/1/commits/4bb4b92163a9f0e704010e5e06ca9cade57094d3

File: test/question-2.js (hunk 1 above)
Comment (rewanthtammana) [reply test/question-2.js:12]: I think you are expecting this kind of test set - https://github.com/rewanth1997/granular-coding-round/pull/1/commits/f6e9556492110e20a95aa53413a5cab36bbe46b5

**Comment by ankuriitk** [comment]
Please give access to @lokeshdevnani  and @Rohanhacker
